solutions/0700-0799/0706-design-hashmap-medium.md

```diff
@@ -0,0 +1,351 @@
+
+As a problem stated, Design HashMap, but what is HashMap or HashTable ?
+
+_A hash tables is a data structure that maps keys to values_ for highly efficient lookup (constant $O(1)$ time).
+
+In this simple implementation, we use an array of linked lists and a hash code function to store the values.
+
+**To Insert a key**,
+
+* First, compute the key’s hash code, which will usually be an `int` or `long`.
+* Then, map the hash code to an index in the array. Eg: `hash(key) % array_length`.
+* _Store the key and value in this index_, we must use linked list because of collisions.
+
+<br />
+
+How can we design the HashMap which is similar to used in programming languages?
+
+Few points to consider while designing the HashMap, 
+
+* How to handle the collisions? Collision is if the same key is mapped to more than value. 
+* HashMap should be efficient and should automatically grow and shrink in size. 
+
+We will look at first, How to handle collisions. There are multiple ways of handling HashTable (HashMap) collision, here I have used Chaining with Linked Lists.
+
+<br />
+
+**Chaining with Linked Lists**
+
+The hash table’s array maps to a linked list of items. We just add the items to this linked list. As long as the number of collisions is fairly small, this will be quite efficient.
+
+In the worst case, lookup is $O(n)$, where the $n$ is the number of elements in the Linked List.
+
+<br />
+
+**Automatically increase and decrease in size**
+
+Since internally, all `HashMap` implementation uses _array to store the values_. 
+
+As we know array's can't be increase or decrease in size automatically, because once we initialize the array with a fixed size, the array is default to the size, until we re-assign the values with new array or re-initialize the array with the new size when program needs. 
+
+But HashMap stores and operates thousands and millions of elements seamlessly with constant $O(1)$ time. As technically it's `amortized` constant $O(1)$ time. 
+
+What is amortized time?
+
+When the array in it has reached its capacity and need to re-create itself with the doubled size. The array has hit the capacity and we have no slots available. Then we need to create a brand new array with the doubled size. And then copy the items in the old one to the new one, which takes $O(n)$ where $n$ is the capacity of the old array and the worst case.
+
+<br />
+
+The implementation of HashMap increase / decrease in size happens in the below cases.
+
+Initial Capacity is $16$ and I have used `Load Factor` as $75$%, As soon as the size reaches $75$%, Hashmap does table doubling and shrink automatically ($~30$%).
+
+If $m > n$, HashMap allocating the memory for $2*N$ elements and **rehasing the entire array for the newly allocated slots**.
+
+If $n / 4$ size reached while removing elements, then Hashmap shrink it to $n / 2$ size.
+
+<br />
+
+<Tabs>
+<TabItem value="java" label="Java">
+<SolutionAuthor name="@vigneshshiv"/>
+
+```java
+class MyHashMap {
```

Comment:
the code is a bit complex for beginners and not fully explain in above content. suggest to add more comments for each function or even line by line.

Reply:
Comments added, check if this is Ok. 

Since it requires basic understanding of Linked List and HashMap internal working. Tried my best to elaborate, check if this Ok